File: ios/Releaf/Features/Notebook/ShelfDetailViewModel.swift
/*
 * ShelfDetailViewModel.swift
 *
 * Room/GRDB-backed VM for the variant-1 chapters screen. Observes
 * the notebook + its chapters + its pages and re-groups into the
 * domain `NotebookDetail` shape the UI already speaks. Display-only
 * fields (`shelfName`, `volumeNumber`, `iconKey`, `colorToken`) are
 * derived here from what the schema actually persists so the UI
 * doesn't carry duplicate mapping logic.
 */

import Foundation
import ReleafData

@MainActor
public final class ShelfDetailViewModel: ObservableObject {

    public enum State: Equatable {
        case loading
        case loaded(NotebookDetail)
        case failed(String)
    }

    @Published public private(set) var state: State = .loading

    private let notebookId: String
    private let notebookRepository: NotebookRepository
    private let chapterRepository: ChapterRepository
    private let pageRepository: PageRepository

    private var tasks: [Task<Void, Never>] = []
    private var latestNotebook: NotebookEntity?
    private var latestChapters: [ChapterEntity] = []
    private var latestPages: [PageEntity] = []

    public init(
        notebookId: String,
        notebookRepository: NotebookRepository = NotebookRepository(),
        chapterRepository: ChapterRepository = ChapterRepository(),
        pageRepository: PageRepository = PageRepository()
    ) {
        self.notebookId = notebookId
        self.notebookRepository = notebookRepository
        self.chapterRepository = chapterRepository
        self.pageRepository = pageRepository
    }

    public func start() {
        stop()

        tasks.append(Task { [weak self, notebookRepository, notebookId] in
            guard let self else { return }
            do {
                for try await value in notebookRepository.observeById(notebookId) {
                    self.latestNotebook = value
                    self.recompute()
                }
            } catch {}
        })
        tasks.append(Task { [weak self, chapterRepository, notebookId] in
            guard let self else { return }
            do {
                for try await value in chapterRepository.observeForNotebook(notebookId: notebookId) {
                    self.latestChapters = value
                    self.recompute()
                }
            } catch {}
        })
        tasks.append(Task { [weak self, pageRepository, notebookId] in
            guard let self else { return }
            do {
                for try await value in pageRepository.observeForNotebook(notebookId: notebookId) {
                    self.latestPages = value
                    self.recompute()
                }
            } catch {}
        })
    }

    public func stop() {
        tasks.forEach { $0.cancel() }
        tasks.removeAll()
    }

    deinit { tasks.forEach { $0.cancel() } }

    // MARK: - Actions

    public func createChapter(title: String = "") {
        let resolved = title.trimmingCharacters(in: .whitespacesAndNewlines)
            .ifEmpty("Untitled chapter")
        Task { [chapterRepository, notebookId] in
            _ = try? await chapterRepository.createChapter(notebookId: notebookId, title: resolved)
        }
    }

    public func createPage(in chapterId: String, onCreated: @escaping (String) -> Void = { _ in }) {
        Task { [pageRepository] in
            if let created = try? await pageRepository.createPage(
                chapterId: chapterId, title: nil, notes: ""
            ) {
                await MainActor.run { onCreated(created.id) }
            }
        }
    }

    /// Add a new volume of this book. Promotes the book into a
    /// series on first tap (via `ensureSeriesFor`) and appends the
    /// next volume. `volumeName` is optional — leave nil to let the
    /// repository render "<series> vol <n>".
    public func addVolume(
        volumeName: String? = nil,
        onCreated: @escaping (String) -> Void = { _ in }
    ) {
        Task { [notebookRepository, notebookId] in
            do {
                let seriesId = try await notebookRepository.ensureSeriesFor(notebookId: notebookId)
                let created = try await notebookRepository.addVolumeToSeries(
                    seriesId:   seriesId,
                    volumeName: volumeName
                )
                await MainActor.run { onCreated(created.id) }
            } catch {
                // Swallow — UI will remain on the current book; the
                // caller sees nothing change, which is the same
                // no-op behavior as other failed writes today.
            }
        }
    }

    // MARK: - Recompute

    private func recompute() {
        guard let entity = latestNotebook else {
            state = latestChapters.isEmpty && latestPages.isEmpty
                ? .loading
                : .failed("Notebook not found")
            return
        }

        let pagesByChapter = Dictionary(grouping: latestPages, by: { $0.chapterId })
        let chapters: [Chapter] = latestChapters.map { chEntity in
            let pages = (pagesByChapter[chEntity.id] ?? [])
                .sorted { a, b in
                    if a.position != b.position { return a.position < b.position }
                    return a.createdAt < b.createdAt
                }
                .map(ShelfMapper.toPageSummary)
            return Chapter(
                id:         chEntity.id,
                notebookId: chEntity.notebookId,
                title:      chEntity.title,
                position:   max(Int(chEntity.position / 1024), 0),
                updatedAt:  ShelfMapper.parseISO(chEntity.updatedAt) ?? Date(timeIntervalSince1970: 0),
                pages:      pages
            )
        }

        let notebook = ShelfMapper.toNotebook(
            entity: entity,
            chapterCount: chapters.count,
            pageCount: latestPages.count
        )

        state = .loaded(NotebookDetail(notebook: notebook, chapters: chapters))
    }
}

// MARK: - Shared entity → domain mapper

enum ShelfMapper {

    static func toNotebook(
        entity: NotebookEntity,
        chapterCount: Int,
        pageCount: Int
    ) -> Notebook {
        let token = colorToken(for: entity.colorHex)
        return Notebook(
            id:           entity.id,
            title:        entity.title,
            description:  nil,
            colorToken:   token,
            position:     max(Int(entity.position / 1024), 0),
            archivedAt:   nil,
            updatedAt:    parseISO(entity.updatedAt) ?? Date(timeIntervalSince1970: 0),
            chapterCount: chapterCount,
            pageCount:    pageCount,
            shelfName:    shelfName(from: entity.title),
            volumeNumber: max(Int(entity.position / 1024), 1),
            status:       .active,
            iconKey:      iconKey(for: token)
        )
    }

    static func toPageSummary(_ entity: PageEntity) -> PageSummary {
        let attachments = entity.attachments.parseAttachments()
        let photos = attachments.filter { $0.type == Attachment.typePhoto }.count
        let scans  = attachments.filter { $0.type == Attachment.typeScan  }.count
        let voice  = attachments.filter { $0.type == Attachment.typeVoice }.count
        return PageSummary(
            id:         entity.id,
            title:      entity.title?.nonEmpty ?? "Untitled page",
            capturedOn: humanDate(entity.createdAt),
            updatedAt:  parseISO(entity.updatedAt) ?? Date(timeIntervalSince1970: 0),
            counts: PageCounts(
                photos: photos,
                voiceNotes: voice,
                scannedDocuments: scans
            ),
            tags: []
        )
    }

    static func colorToken(for hex: String?) -> String {
        let cleaned = hex?.uppercased().replacingOccurrences(of: "#", with: "")
        switch cleaned {
        case "7AA874": return "green"
        case "E07856": return "coral"
        case "F4C430": return "yellow"
        case "B8956A": return "dry"
        case "8E86DB", "6E66BB": return "info"
        default: return "green"
        }
    }

    static func iconKey(for colorToken: String) -> String {
        switch colorToken {
        case "green":  return "plant"
        case "info":   return "chart"
        case "dry":    return "sun"
        case "coral":  return "book"
        case "yellow": return "sun"
        default:       return "plant"
        }
    }

    static func shelfName(from title: String) -> String {
        let trimmed = title.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return "NOTEBOOK" }
        let head = trimmed
            .split(separator: " ", maxSplits: 1, omittingEmptySubsequences: true)
            .first.map(String.init) ?? trimmed
        let clean = head.split(separator: "—").first.map(String.init) ?? head
        let upper = clean.uppercased()
        return upper.isEmpty ? "NOTEBOOK" : upper
    }

    static func parseISO(_ s: String) -> Date? {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = f.date(from: s) { return d }
        f.formatOptions = [.withInternetDateTime]
        return f.date(from: s)
    }

    static func humanDate(_ s: String) -> String? {
        guard let date = parseISO(s) else { return nil }
        let f = DateFormatter()
        f.dateFormat = "MMM d, yyyy"
        return f.string(from: date)
    }
}

private extension String {
    var nonEmpty: String? { trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : self }
    func ifEmpty(_ fallback: String) -> String {
        trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? fallback : self
    }
}

File: android/app/src/main/java/app/releaf/mobile/features/notebook/ShelfDetailViewModel.kt
/*
 * ShelfDetailViewModel.kt
 *
 * Room-backed VM for the variant-1 chapters screen. Observes the
 * notebook metadata, its chapters, and the flat list of every page
 * beneath, then regroups into `(Notebook, List<Chapter with pages>)`
 * in the domain shape the variant UI already speaks.
 *
 * Display-only fields (`shelfName`, `volumeNumber`, `iconKey`,
 * `colorToken`) are derived here from what the schema actually
 * persists (title, position, color_hex) so the UI doesn't carry
 * duplicate mapping logic.
 */

package app.releaf.mobile.features.notebook

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.createSavedStateHandle
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import app.releaf.mobile.ReleafApp
import app.releaf.mobile.data.domain.Chapter
import app.releaf.mobile.data.domain.Notebook
import app.releaf.mobile.data.domain.NotebookStatus
import app.releaf.mobile.data.domain.PageCounts
import app.releaf.mobile.data.domain.PageSummary
import app.releaf.mobile.data.drive.NotebookDetail
import app.releaf.mobile.data.notebook.ChapterRepository
import app.releaf.mobile.data.notebook.NotebookRepository
import app.releaf.mobile.data.notebook.PageEntity
import app.releaf.mobile.data.notebook.PageRepository
import app.releaf.mobile.data.notebook.parseAttachments
import java.time.Instant
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

sealed interface ShelfDetailUiState {
    data object Loading : ShelfDetailUiState
    data class Loaded(val detail: NotebookDetail) : ShelfDetailUiState
    data class Failed(val message: String) : ShelfDetailUiState
}

class ShelfDetailViewModel(
    application: Application,
    private val notebookId: String,
    private val notebookRepository: NotebookRepository,
    private val chapterRepository: ChapterRepository,
    private val pageRepository: PageRepository,
) : AndroidViewModel(application) {

    val state: StateFlow<ShelfDetailUiState> = combine(
        notebookRepository.observeById(notebookId),
        chapterRepository.observeForNotebook(notebookId),
        pageRepository.observeForNotebook(notebookId),
    ) { notebook, chapters, pages ->
        if (notebook == null) {
            ShelfDetailUiState.Failed("Notebook not found") as ShelfDetailUiState
        } else {
            val pagesByChapter = pages.groupBy { it.chapterId }
            val domainChapters = chapters.map { entity ->
                val chapterPages = (pagesByChapter[entity.id].orEmpty())
                    .sortedWith(compareBy({ it.position }, { it.createdAt }))
                Chapter(
                    id         = entity.id,
                    notebookId = entity.notebookId,
                    title      = entity.title,
                    position   = (entity.position / 1024L).toInt().coerceAtLeast(0),
                    updatedAt  = parseIso(entity.updatedAt),
                    pages      = chapterPages.map(::toPageSummary),
                )
            }
            val nb = notebook.toDomain(
                chapterCount = domainChapters.size,
                pageCount    = pages.size,
            )
            ShelfDetailUiState.Loaded(NotebookDetail(nb, domainChapters)) as ShelfDetailUiState
        }
    }.stateIn(
        scope        = viewModelScope,
        started      = SharingStarted.WhileSubscribed(5_000),
        initialValue = ShelfDetailUiState.Loading,
    )

    /** Create an empty chapter. Shelves UI will re-render from the flow. */
    fun createChapter(
        title: String = "",
        onCreated: (String) -> Unit = {},
    ) {
        val resolved = title.trim().ifEmpty { "Untitled chapter" }
        viewModelScope.launch {
            val created = chapterRepository.createChapter(notebookId, resolved)
            onCreated(created.id)
        }
    }

    /** Create an empty page under [chapterId]; used when a chapter row has
     *  no pages yet and the user taps it from the shelves chapters view. */
    fun createPage(chapterId: String, onCreated: (String) -> Unit = {}) {
        viewModelScope.launch {
            val created = pageRepository.createPage(
                chapterId = chapterId,
                title     = null,
                notes     = "",
            )
            onCreated(created.id)
        }
    }

    /**
     * Add a new volume of this book. If the book isn't part of a
     * series yet, promote it into one (using the book's current
     * title as the series name) and then append Volume 2. Returns
     * the new volume's id via [onCreated] so the caller can
     * navigate straight into it.
     *
     * `volumeName` is optional — leave blank to let the repository
     * render the default "<series> vol <n>" label.
     */
    fun addVolume(
        volumeName: String? = null,
        onCreated: (String) -> Unit = {},
    ) {
        viewModelScope.launch {
            val seriesId = notebookRepository.ensureSeriesFor(notebookId = notebookId)
            val created = notebookRepository.addVolumeToSeries(
                seriesId   = seriesId,
                volumeName = volumeName,
            )
            onCreated(created.id)
        }
    }

    companion object {
        const val ARG_NOTEBOOK_ID = "notebookId"

        val Factory: ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val app = this[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY] as ReleafApp
                val saved: SavedStateHandle = createSavedStateHandle()
                val notebookId = checkNotNull(saved.get<String>(ARG_NOTEBOOK_ID)) {
                    "ShelfDetailViewModel missing $ARG_NOTEBOOK_ID"
                }
                ShelfDetailViewModel(
                    application        = app,
                    notebookId         = notebookId,
                    notebookRepository = app.notebookRepository,
                    chapterRepository  = app.chapterRepository,
                    pageRepository     = app.pageRepository,
                )
            }
        }
    }
}

// ---------- entity → domain mapping ----------

private fun app.releaf.mobile.data.notebook.NotebookEntity.toDomain(
    chapterCount: Int,
    pageCount: Int,
): Notebook {
    val token = shelfColorToken(colorHex)
    return Notebook(
        id           = id,
        title        = title,
        description  = null,
        colorToken   = token,
        position     = (position / 1024L).toInt().coerceAtLeast(0),
        archivedAt   = null,
        updatedAt    = parseIso(updatedAt),
        chapterCount = chapterCount,
        pageCount    = pageCount,
        shelfName    = shelfNameFor(title),
        volumeNumber = (position / 1024L).toInt().coerceAtLeast(1),
        status       = NotebookStatus.Active,
        iconKey      = iconKeyFor(token),
    )
}

private fun toPageSummary(entity: PageEntity): PageSummary {
    val attachments = runCatching { entity.attachments.parseAttachments() }
        .getOrDefault(emptyList())
    val photos = attachments.count { it.type == "photo" }
    val scans  = attachments.count { it.type == "scan" }
    val voice  = attachments.count { it.type == "voice" }
    return PageSummary(
        id         = entity.id,
        title      = entity.title?.takeIf { it.isNotBlank() } ?: "Untitled page",
        capturedOn = humanDate(entity.createdAt),
        updatedAt  = parseIso(entity.updatedAt),
        counts     = PageCounts(
            photos           = photos,
            voiceNotes       = voice,
            scannedDocuments = scans,
        ),
        // No tag column yet; the variant page view derives its own
        // display tags. Left empty so callers don't treat fake tags
        // as real.
        tags       = emptyList(),
    )
}

private fun shelfNameFor(title: String): String {
    val t = title.trim()
    if (t.isEmpty()) return "NOTEBOOK"
    val head = t.substringBefore(' ').substringBefore('—').trim()
    return head.uppercase().ifEmpty { "NOTEBOOK" }
}

private fun shelfColorToken(hex: String?): String = when (hex?.uppercase()?.removePrefix("#")) {
    "7AA874" -> "green"
    "E07856" -> "coral"
    "F4C430" -> "yellow"
    "B8956A" -> "dry"
    "8E86DB", "6E66BB" -> "info"
    else -> "green"
}

private fun iconKeyFor(colorToken: String): String = when (colorToken) {
    "green"  -> "plant"
    "info"   -> "chart"
    "dry"    -> "sun"
    "coral"  -> "book"
    "yellow" -> "sun"
    else     -> "plant"
}

private fun parseIso(iso: String): Instant =
    runCatching { Instant.parse(iso) }.getOrDefault(Instant.EPOCH)

private fun humanDate(iso: String): String {
    val instant = runCatching { Instant.parse(iso) }.getOrNull() ?: return ""
    val fmt = java.time.format.DateTimeFormatter.ofPattern("MMM d, yyyy")
        .withZone(java.time.ZoneId.systemDefault())
    return fmt.format(instant)
}
